src/CodeGen_C.cpp

```diff
@@ -181,7 +179,38 @@ const string preamble =
     "}\n";
 }
 
-CodeGen_C::CodeGen_C(ostream &s) : IRPrinter(s), id("$$ BAD ID $$") {}
+CodeGen_C::CodeGen_C(ostream &s, bool is_header, const std::string &guard) : IRPrinter(s), id("$$ BAD ID $$"), is_header(is_header) {
+    if (is_header) {
+        // If it's a header, emit an include guard.
+        stream << "#ifndef HALIDE_" << print_name(guard) << '\n'
+               << "#define HALIDE_" << print_name(guard) << '\n';
+    }
+
+    stream << "#ifdef __cplusplus\n";
+    stream << "#define HALIDE_EXTERN_C extern \"C\"\n";
+    stream << "#else\n";
+    stream << "#define HALIDE_EXTERN_C\n";
```

Comment:
I don't think this would work, the macro gets used even on function definitions. I think it might be an error to do:

    extern void f() { ... }

?

Either way, I don't think it will change anything. 'extern' is only necessary to get the behavior defined by extern "C".